Fix apy modules tests - Re: #516

tomasff opened the pull request:
This pull request attempts to fix issue #516.

Two changes were made:
1. In the apy module a specific value stored in the constant `SYNTAX_ERROR` is returned if a syntax error occurred. This change was made so that several functions in this module can be tested again. (Related to changes in #504)
2. A new function `check_syntax_error` was created to test if a function returns the appropriate `SYNTAX_ERROR` code.

Althought this is not a very elegant solution it should allow us to test for any problems in this module when changes occur, while a more robust solution with maybe more specific exceptions isn't implemented.

tomasff committed:
Return value if a syntax error has occurred.

tomasff committed:
Fixed missing return SYNTAX_ERROR in function

tomasff committed:
Fixed yet another missing return of SYNTAX_ERROR

tomasff committed:
Update apy module tests to check for syntax errors

A new function check_syntax_error was created to test if the appropriate syntax error is returned.

Androbin commented:
Maybe we could differentiate between `LINE_ERROR`, `CMD_ERROR`, `TEXT_ERROR` and so on?
LGTM otherwise :+1:

tomasff commented:
@Androbin Sure, that's fine. Should I maybe create an enum or a dictionary?
The dictionary could be as follows:
```python
errors = {
    'SYNTAX_ERROR': {
        'LINE_ERROR': 1,
        'CMD_ERROR': 2,
        'TEXT_ERROR': 3
    }
}
```
Or with an enum:
```python
class SyntaxError(Enum):
    	LINE_ERROR = 1
        CMD_ERROR = 2
        TEXT_ERROR = 3
```
I could also maintain the current structure of adding constants at the top. However the dictionary seems more flexible, in my opinion. Could you please let me know which solution would be best?

Androbin commented:
Nice writeup! I would argue for the `enum` variant. I think this best conveys the idea of getting exactly one of a known set of return values. Actually, I think we're also missing an `OK` value or similar.

Androbin commented:
I do have to say, this actually looks quite suspicious:
```
text, text_error = strict_check(r'.*', input.group(1), apertium_identlang)
```
Actually, after closer inspection, I can't tell why these error names were picked originally.

Androbin commented:
How about `KEYWORD_ERROR` for unknown sub-commands or similar, `ENUM_ERROR` for unknown source/target language or similar, and `SYNTAX_ERROR` for something that is closer to syntax errors in the classic sense like bad delimiters and stuff?

Androbin commented:
For example, a `strict_check` on `pairRE = langRE + '-' + langRE` might lead to either `SYNTAX_ERROR` if there's not exactly one dash, or to `ENUM_ERROR` in case the language is unknown.

tomasff committed:
Created enum for functions status

Created an enum for functions status.
Renamed error variables to a more appropriate name.

tomasff commented:
@Androbin is this what you meant?

wei2912 commented on modules/apy.py:
```diff
@@ -24,6 +25,13 @@
 
 SYNTAX_ERROR = 1
```
What's this line still doing here? :P

tomasff replied:
Forgot to remove it haha. Should be removdd now.

wei2912 commented on modules/test/test_apy.py:
```diff
@@ -55,13 +55,13 @@ def check_exceptions(self, bad_list, name, reason=None):
             self.input.group.return_value = inp
             function_return_val = name.__call__(self.phenny, self.input)
-            self.assertEqual(function_return_val, apy.SYNTAX_ERROR, msg='No syntax error returned for {:s}!'.format(reason))
+            self.assertEqual(function_return_val, error, msg='No syntax error returned for {:s}!'.format(reason))
```
It indicates specifically "syntax error" here, which doesn't seem that general... I can't think of a more elegant solution though. Perhaps we could just restrict the function to syntax errors, as it was previously.

tomasff replied:
@wei2912 I have just pushed a solution for this. Let me know what you think.

wei2912 reviewed (commented):
I suppose there's more work to be done to integrate the other error types into the rest of the code? :P I only see `SYNTAX_ERROR` and `WEB_RESPONSE_ERROR` being used.

tomasff committed:
Remove SYNTAX_ERROR constant which is no longer necessary

tomasff committed:
Display appropriate error message on apy tests

tomasff committed:
Change error type to a more appropriate one

tomasff commented:
@wei2912 I can integrate the other error types, however this would envolve removing a lot of the thrown `GrumbleError`. Is that alright?

tomasff commented:
@wei2912 @Androbin 
Update: I tried integrating other errors types by replacing `GrumbleError`'s with the new `Status` codes.

Unfortunately this broke some of the tests and I wasn't able to fix them. 

In my opinion a good separate issue/ticket would be to think about a solid structure for error handling thats consistent across all modules, and temporarily fix the apy module's tests, so that we can verify if new changes break the module.

Please let me know if I should push my attempt to replace `GrumbleError`'s with the new `Status` codes.

Androbin commented:
The APY tests are green now. I agree with @tomasff
@wei2912 We should merge this now and have him submit another PR.

wei2912 commented:
IMO I don't think it's necessary to standardise the error reporting across all the modules. In nearly any other codebase I'd agree with the need for consistent error handling, but for a modular scripts system that Begiak has, I don't think it'd lead to much payoff. We could have each module handle exceptions as best as they can (since each module would have their own exceptions to deal with), and if they're unable to do so, the bot will attempt to catch the exception and report it unless it leads to an unrecoverable error.

As for replacing `GrumbleError` with the new `Status` codes, are you speaking of modifying other modules? That can be submitted in a separate GCI task and PR.

This is related to https://github.com/apertium/phenny/issues/509.

wei2912 commented:
I'll be putting this into the commit message, let me know if it summarises the details well:
1. An enum was created to deal with different types of errors within the module, so that the tests can be fixed. (Related to changes in https://github.com/apertium/phenny/pull/504.)
2. A new function `check_error` was created to test exception handling.

wei2912 commented on modules/test/test_apy.py:
```diff
@@ -250,5 +261,6 @@ def test_perword(self, mock_open):
         # bad input
         self.check_exceptions(['fra (tagger nonfunc) word'], apy.apertium_perword,
```
Saw that the `check_exceptions` function is still being used here. Can we replace the function completely with `check_error`?

tomasff replied:
@wei2912 I have replaced it with `check_error` 👍 
I also removed `check_exceptions` as it is no longer used.

tomasff committed:
Completly replace check_exceptions with check_error

tomasff commented:
@wei2912 I meant replacing the use of `GrumbleError` in `handle_error` and `check_no_data`. I tried doing this, but a lot of tests related to the `translate` function broke and I wasn't able to fix them. I will try again today. Should I create an issue for this?

I think a standerdise way of handling errors would be benefitial because as of right now, errors are all handled a bit differently in every module: we're using the `Status` codes in the `apy` module and `GrumbleError`'s in other modules.

wei2912 commented:
@tomasff ah, I see. Please do try to replace `GrumbleError` completely and
let us know which tests break. I'll still accept your GCI submission for
fixing the tests, but will not merge until this is done.

How would the `Status` codes generalise to other modules? If you can
propose how to shift towards a standardised way, please feel free to create
an issue for it and I'll write up a few GCI tasks for that. Thanks for the
work so far!

On Sun, 22 Dec 2019, 21:21 Tomás F., <notifications@github.com> wrote:

> @wei2912 <https://github.com/wei2912> I meant replacing the use of
> GrumbleError in handle_error and check_no_data. I tried doing this, but a
> lot of tests related to the translate function broke and I wasn't able to
> fix them. I will try again today. Should I create an issue for this?
>
> I think a standerdise way of handling errors would be benefitial because
> as of right now, errors are all handled a bit differently in every module:
> we're using the Status codes in the apy module and GrumbleError's in
> other modules.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/phenny/pull/522?email_source=notifications&email_token=AAPZPF6OEQLITTRBHSLC7XDQZ5SULA5CNFSM4J6CN6LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHPPV2A#issuecomment-568261352>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPZPF2LT5S35DVDJ6TOCKTQZ5SULANCNFSM4J6CN6LA>
> .
>

wei2912 commented:
Also, no need to create a new issue for the tests failing when you replace
`GrumbleError` with `Status` in this module. Just keep us updated here on
the issues you face.

On Mon, 23 Dec 2019, 06:24 Ng Wei En, <weien1292@gmail.com> wrote:

> @tomasff ah, I see. Please do try to replace `GrumbleError` completely and
> let us know which tests break. I'll still accept your GCI submission for
> fixing the tests, but will not merge until this is done.
>
> How would the `Status` codes generalise to other modules? If you can
> propose how to shift towards a standardised way, please feel free to create
> an issue for it and I'll write up a few GCI tasks for that. Thanks for the
> work so far!
>
> On Sun, 22 Dec 2019, 21:21 Tomás F., <notifications@github.com> wrote:
>
>> @wei2912 <https://github.com/wei2912> I meant replacing the use of
>> GrumbleError in handle_error and check_no_data. I tried doing this, but
>> a lot of tests related to the translate function broke and I wasn't able
>> to fix them. I will try again today. Should I create an issue for this?
>>
>> I think a standerdise way of handling errors would be benefitial because
>> as of right now, errors are all handled a bit differently in every module:
>> we're using the Status codes in the apy module and GrumbleError's in
>> other modules.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/apertium/phenny/pull/522?email_source=notifications&email_token=AAPZPF6OEQLITTRBHSLC7XDQZ5SULA5CNFSM4J6CN6LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHPPV2A#issuecomment-568261352>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAPZPF2LT5S35DVDJ6TOCKTQZ5SULANCNFSM4J6CN6LA>
>> .
>>
>

tomasff commented:
Sorry for the late reply.

I will try to look into this tomorrow.

On Sun, 22 Dec 2019, 22:30 Ng Wei En, <notifications@github.com> wrote:

> Also, no need to create a new issue for the tests failing when you replace
> `GrumbleError` with `Status` in this module. Just keep us updated here on
> the issues you face.
>
> On Mon, 23 Dec 2019, 06:24 Ng Wei En, <weien1292@gmail.com> wrote:
>
> > @tomasff ah, I see. Please do try to replace `GrumbleError` completely
> and
> > let us know which tests break. I'll still accept your GCI submission for
> > fixing the tests, but will not merge until this is done.
> >
> > How would the `Status` codes generalise to other modules? If you can
> > propose how to shift towards a standardised way, please feel free to
> create
> > an issue for it and I'll write up a few GCI tasks for that. Thanks for
> the
> > work so far!
> >
> > On Sun, 22 Dec 2019, 21:21 Tomás F., <notifications@github.com> wrote:
> >
> >> @wei2912 <https://github.com/wei2912> I meant replacing the use of
> >> GrumbleError in handle_error and check_no_data. I tried doing this, but
> >> a lot of tests related to the translate function broke and I wasn't able
> >> to fix them. I will try again today. Should I create an issue for this?
> >>
> >> I think a standerdise way of handling errors would be benefitial because
> >> as of right now, errors are all handled a bit differently in every
> module:
> >> we're using the Status codes in the apy module and GrumbleError's in
> >> other modules.
> >>
> >> —
> >> You are receiving this because you were mentioned.
> >> Reply to this email directly, view it on GitHub
> >> <
> https://github.com/apertium/phenny/pull/522?email_source=notifications&email_token=AAPZPF6OEQLITTRBHSLC7XDQZ5SULA5CNFSM4J6CN6LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHPPV2A#issuecomment-568261352
> >,
> >> or unsubscribe
> >> <
> https://github.com/notifications/unsubscribe-auth/AAPZPF2LT5S35DVDJ6TOCKTQZ5SULANCNFSM4J6CN6LA
> >
> >> .
> >>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/phenny/pull/522?email_source=notifications&email_token=ADTBQKBBL47NIYSP6IZKBGDQZ7TARA5CNFSM4J6CN6LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHP3FII#issuecomment-568308385>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADTBQKGME6OWX4XJNA5W2FTQZ7TARANCNFSM4J6CN6LA>
> .
>

wei2912 commented:
@tomasff have there been any updates ever since your last comment? :P

tomasff commented:
Hi,

Not yet. I've been busy with Christmas. Hopefully I'll have it ready for
tomorrow.
So sorry for the delay

On Sat, 28 Dec 2019 at 15:58, Ng Wei En <notifications@github.com> wrote:

> @tomasff <https://github.com/tomasff> have there been any updates ever
> since your last comment? :P
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/phenny/pull/522?email_source=notifications&email_token=ADTBQKGSRQMNH5ZTWUKBXEDQ25ZSRA5CNFSM4J6CN6LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHYMRBQ#issuecomment-569428102>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADTBQKC2OEWOSG2WG3LRIPTQ25ZSRANCNFSM4J6CN6LA>
> .
>

wei2912 commented:
@tomasff Hi, I just realised this PR hasn't been merged yet... how's the progress on this? :P

tomasff commented:
@wei2912 Sorry for the delay, I've been busy these past weeks. I'll try to finish it by next week.

wei2912 commented:
No worries, please take your time! It's not an urgent matter.

On Fri, Feb 21, 2020 at 11:24 PM Tomás F. <notifications@github.com> wrote:

> @wei2912 <https://github.com/wei2912> Sorry for the delay, I've been busy
> these past weeks. I'll try to finish it by next week.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/phenny/pull/522?email_source=notifications&email_token=AAPZPFYE3M5MDEVHLRP6DYLRD7W35A5CNFSM4J6CN6LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMTBZHI#issuecomment-589700253>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPZPF7GO66TAPJOEZQ3TH3RD7W35ANCNFSM4J6CN6LA>
> .
>
